File: problems/14_2d_dp/0010_regular_expression_matching/solution.dart
// 0010. Regular Expression Matching
bool isMatch(String s, String p) {
  final m = s.length, n = p.length;
  final dp = List.generate(m + 1, (_) => List<bool>.filled(n + 1, false));
  dp[0][0] = true;
  for (var j = 1; j <= n; j++) {
    if (p[j - 1] == '*') dp[0][j] = dp[0][j - 2];
  }
  for (var i = 1; i <= m; i++) {
    for (var j = 1; j <= n; j++) {
      if (p[j - 1] == '*') {
        dp[i][j] = dp[i][j - 2];
        if (p[j - 2] == '.' || p[j - 2] == s[i - 1]) {
          dp[i][j] = dp[i][j] || dp[i - 1][j];
        }
      } else if (p[j - 1] == '.' || p[j - 1] == s[i - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      }
    }
  }
  return dp[m][n];
}
void main() {
  if (isMatch("aa", "a") != false) throw 'a';
  if (isMatch("aa", "a*") != true) throw 'b';
  if (isMatch("ab", ".*") != true) throw 'c';
  if (isMatch("aab", "c*a*b") != true) throw 'd';
  if (isMatch("mississippi", "mis*is*p*.") != false) throw 'e';
  if (isMatch("", "") != true) throw 'f';
  if (isMatch("", ".*") != true) throw 'g';
  if (isMatch("a", "ab*") != true) throw 'h';
  print('All tests passed!');
}

File: problems/14_2d_dp/0010_regular_expression_matching/solution.py
"""
Problem: 0010. Regular Expression Matching
"""


class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False] * (n + 1) for _ in range(m + 1)]
        dp[0][0] = True
        for j in range(1, n + 1):
            if p[j - 1] == '*':
                dp[0][j] = dp[0][j - 2]
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if p[j - 1] == '*':
                    dp[i][j] = dp[i][j - 2]
                    if p[j - 2] == '.' or p[j - 2] == s[i - 1]:
                        dp[i][j] = dp[i][j] or dp[i - 1][j]
                elif p[j - 1] == '.' or p[j - 1] == s[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
        return dp[m][n]


if __name__ == "__main__":
    sol = Solution()
    assert sol.isMatch("aa", "a") is False
    assert sol.isMatch("aa", "a*") is True
    assert sol.isMatch("ab", ".*") is True
    assert sol.isMatch("aab", "c*a*b") is True
    assert sol.isMatch("mississippi", "mis*is*p*.") is False
    assert sol.isMatch("", "") is True
    assert sol.isMatch("", ".*") is True
    assert sol.isMatch("a", "ab*") is True
    print("All tests passed!")
